src/firmware/linux/host/types.rs

```diff
@@ -285,61 +284,83 @@ pub struct CertTableEntry {
     length: u32,
 }
 
-#[repr(C)]
-pub struct CertTable {
-    /// Pointer to the Certificates in question.
-    entries: *const CertTableEntry,
-}
+impl CertTableEntry {
+    /// Builds a Kernel formatted CertTable for sending the certificate content to the PSP.
+    ///
+    /// Users should pass the rust-friendly vector of [`UapiCertTableEntry`], and this function
+    /// will handle adding the last entry and the structuring of the buffer sent to the hypervisor.
+    ///
+    /// The contiguous memory layout should look similar to this:
+    ///
+    /// ```text
+    ///             |-> |------------------|    |-  CertTableEntry -|
+    ///             |   | CertTableEntry_1 <<<--| - guid            |
+    ///             |   | CertTableEntry_2 |    | - offset          |
+    /// CertTable --|   | ...              |    | - length          |
+    ///             |   | ...              |    |-------------------|
+    ///             |   | ...              |
+    ///             |-> | CertTableEntry_z | <-- last entry all zeroes
+    /// offset (1)  --> | RawCertificate_1 |
+    ///                 | ...              |
+    ///                 | ...              |
+    /// offset (2)  --> | RawCertificate_2 |
+    ///                 | ...              |
+    ///                 | ...              |
+    /// offset (n)  --> | RawCertificate_n |
+    ///                 |------------------|
+    ///          
+    /// ```
+    ///
+    pub fn uapi_to_vec_bytes(table: &mut Vec<UapiCertTableEntry>) -> Result<Vec<u8>, SnpCertError> {
```

Comment:
It would be helpful to have a unit-test for this function and for `parse_table`. They do not rely on any kernel-y/ioctl-y stuff, so should be a bit easier to supply some cert table and convert it to bytes and verify that result, and the other way around for parsing.
(sorry if I missed existing tests somewhere)

Reply:
That is definitely on my todo list. I want to test each of these functions as much as possible; __especially__ where there is no hardware or kernel requirements.

Reply:
I am going to create a separate PR to add all of the unit tests so we aren't delaying the next release.